### ISSUE TITLE
Analytics: visible event reset

### ISSUE THREAD

**Issue by zhouyx** [ampproject/amphtml#11987]
## Overview ##
Amp Analytics support visible trigger. Amp analytics will trigger visible event once and only once during the first time `visibilitySpec` is met. People want more visibility events from the page
`visibilitySpec` looks like. Note: all attributes are optional
```
'visibilitySpec': {
  'visiblePercentageMin': 50,
  'visiblePercentageMax': 80,
  'totalTimeMin': 3000,
  'totalTimeMax': 5000,
  'continousTimeMin': 1000,
  'continousTimeMax': 1500,
}
```

### Reset visible event ###
People want a way to reset visibility state and send identical visible event more than once. 

### `reset` ###
We can introduce a new `reset` keyword, that enable visible state resetting after all conditions met. 
For example
```
'visibilitySpec': {
  'visiblePercentageMin': 50,
  'continousTimeMin': 1000,
  'reset': true
}
```
This would generate one visible analytics event if the element has been visible for more than 50% for a continuous of 1sec. If the element is visible for 3sec, 3 visible analytics event will be sent.
```
'visibilitySpec': {
  'reset': true
}
```
This would work just like an IntersectionObserver with threshold 0. Note that with `reset: true`, we need to treat `continousTimeMin: 0` differently. 

For example
```
'inob': {
  'on': 'visible',
  'selector': ... ,
  'request': 'fake.com/analytics?rect=${intersectionRect}ratio=${intersectionRaio}'
  'visibleSpec': {
    'visiblePercentageIncrement': 5,
    'reset': true,
  }
}
```
The config above can serve like our polyfill IntersectionObserver. 

## Concerns and Restrictions ##
With the new spec. Visible events are kept alive all the time, and it is possible that a single trigger spec can trigger a huge list of visible events. We need some restrictions.

- `reset` can only be set to true by vendor
- `reset` can only be set to true with `visible` trigger. Not `hidden` trigger.
- with `reset: true`, `continousTimeMin` and `totalTimeMin` all need to be greater than 200ms (or some value). Otherwise we should ignore the concept of time, and only fire visible event when visible percentage crosses threshold.

**Comment by zhouyx** [comment]
> `reset` can only be set to true by vendor
We don't really need this

Make make `reset` an interval value so it's more clear

**Comment by lannka** [comment]
> We can introduce a new reset keyword, that enable visible state resetting after all conditions met.

Say we have a trigger with visibleSpec:
```
'visibleSpec': {
    'visiblePercentageMin': 50,
    'reset': true,
  }
```
Imagine the element was 60% visible, which met the spec hence a request was sent. However, 
it's a bit unclear what should happen after resetting the state. The element is still 60% visible, should it trigger again immediately? Apparently not.

Here, resetting the state is not enough for a repeated trigger. It should wait for the condition to be unmet as well.

We can use keyword `repeat` to better describe this behavior:

- when there is duration condition, `repeat` means to have the duration condition is met again. 
- when there is only visible percentage condition but no duration condition, `repeat` means to wait for percentage condition unmet and then met again.

**Comment by lannka** [comment]
`repeat` can take an integer value as time interval. 

```
'visibleSpec': {
    'visiblePercentageMin': 50,
    'repeat': 1000,
  }
```
It translates to: 
- if the element is >50% visible, trigger is fired. 
- the trigger is repeated if the element is <50% and then >50% visible again. (same as `repeat: true`)
- the trigger can also be repeated if the element is continuously >50% visible for 1 second

This interval feature can be used as heartbeats to measure total visible time.

**Comment by jonkeller** [comment]
Note to self: Also trigger when 0% visible

**Comment by jonkeller** [comment]
@lannka I think this sufficiently handles "wait until condition is unmet and then met again"?
https://github.com/ampproject/amphtml/pull/12094/files#diff-61c963752cc0a61e2ec5e1561dc0466cR221

this.waitToRefresh_ is then used here:
https://github.com/ampproject/amphtml/pull/12094/files#diff-61c963752cc0a61e2ec5e1561dc0466cR332

And then the "when there is duration condition..." part is already addressed here, right? 
https://github.com/ampproject/amphtml/pull/12094/files#diff-61c963752cc0a61e2ec5e1561dc0466cR216
(this.resetInterval_ is the max of totalTimeMin, continuousTimeMin, and the numeric value of reset (now changed to repeat).

So, it seems to me like everything you're requested was already done by Yuxuan...unless I am not understanding you correctly?

**Comment by rudygalfi** [comment]
What's the status of this?

**Comment by zhouyx** [comment]
Looks like implementation has been merged and documented. Closing for now. @jonkeller Please correct me if there's any remaining work.

**Comment by jonkeller** [comment]
Agree, nothing remaining. Thanks!
